### PR TITLE
Fix modules reload leak

### LINKED ISSUE #109: Reloading modules leaks memory
Clone [dups](https://github.com/linuxwhatelse/dups).
Open it
Open few files, do some random go to definitions.
Leave product inactive

Observe RAM growth and some odd module reloading

image

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Fixes #109

Simplifies event listening and make things happen in proper order. Reloading modules no longer increases RAM.

### CODE REVIEW

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs [hunk 1]
```diff
@@ -274,6 +274,7 @@ internal class AstPythonInterpreter : IPythonInterpreter2, IModuleContext, ICanF
             }
 
             _analyzer = state;
+            RemoveImportedModules();
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs]: Do we need a lock here?

File: src/Analysis/Engine/Impl/PythonAnalyzer.cs [hunk 2]
```diff
@@ -174,16 +174,16 @@ public partial class PythonAnalyzer : IPythonAnalyzer, IDisposable {
...
                 _interpreter.Initialize(this);
 
                 await LoadKnownTypesAsync(token);
-
                 foreach (var mod in _modulesByFilename.Values) {
                     mod.Clear();
                     mod.EnsureModuleVariables(this);
                 }
+
+                _interpreterFactory.NotifyImportNamesChanged();
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/PythonAnalyzer.cs]: Who is listening to this `ImportableModulesChanged` event?

Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/PythonAnalyzer.cs]: AstFactory и AstIntepreter, они там кэш стирают

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs (hunk 1 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs]: Code locks on `_userSearchPathImported` operations

Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs]: I am resolving it on another work with importing/search paths

**Comment by MikhailArkhipov** [comment]
https://github.com/Microsoft/python-language-server/issues/209 first

**Pull request closed without merging**